File: problems/0801-0900/0843_guess-the-word/provided/go/oracle.go
package main

// Problem-provided oracle (Master), Go side. Compiled beside
// every submission by the judge; never editable in the editor.
// Constructed from the case state: the wordlist and the secret word as
// generic values, then the guess budget.
type Master struct {
	secret string
	found  bool
	budget int64
}

// NewMaster builds the oracle from the case's construction values
// (the wordlist and the secret word) and the guess budget.
func NewMaster(construction []any, budget int64) *Master {
	if _, ok := construction[0].([]any); !ok {
		panic("Master wordlist must be an array")
	}
	secret, ok := construction[1].(string)
	if !ok {
		panic("Master secret must be a string")
	}
	return &Master{secret: secret, budget: budget}
}

// Guess answers the number of positions where word and the secret word
// agree, and records whether the secret itself was named.
func (interrogator *Master) Guess(word string) int {
	if interrogator.budget <= 0 {
		panic("Master guess budget exhausted")
	}
	interrogator.budget--
	if word == interrogator.secret {
		interrogator.found = true
	}
	matches := 0
	n := len(word)
	if len(interrogator.secret) < n {
		n = len(interrogator.secret)
	}
	for i := 0; i < n; i++ {
		if word[i] == interrogator.secret[i] {
			matches++
		}
	}
	return matches
}

// Verdict reports whether the secret word was named within the budget.
func (interrogator *Master) Verdict() any {
	return interrogator.found
}

File: problems/0801-0900/0843_guess-the-word/provided/rust/oracle.rs
// Problem-provided oracle (Master), Rust side. Assembled into
// every submission's crate by the judge; never editable in the editor.
// Constructed from the case state: the wordlist and the secret word as
// generic values, then the guess budget.
#[allow(dead_code)]
pub struct Master {
    secret: String,
    found: bool,
    budget: i64,
}

impl Master {
    pub fn new(construction: &[OjValue], budget: i64) -> Self {
        match construction.first() {
            Some(OjValue::Array(_)) => {}
            _ => panic!("Master wordlist must be an array"),
        }
        let secret = match construction.get(1) {
            Some(OjValue::Str(text)) => text.clone(),
            _ => panic!("Master secret must be a string"),
        };
        Master {
            secret,
            found: false,
            budget,
        }
    }

    pub fn guess(&mut self, word: &str) -> i32 {
        if self.budget <= 0 {
            panic!("Master guess budget exhausted");
        }
        self.budget -= 1;
        if word == self.secret {
            self.found = true;
        }
        self.secret.bytes().zip(word.bytes()).filter(|(a, b)| a == b).count() as i32
    }

    pub fn verdict(&self) -> OjValue {
        OjValue::Bool(self.found)
    }
}
